sycl/include/sycl/detail/impl_utils.hpp

```diff
@@ -0,0 +1,55 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/detail/defines_elementary.hpp>
+
+#include <cassert>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace detail {
+
+// Helper function for extracting implementation from SYCL's interface objects.
+// Note! This function relies on the fact that all SYCL interface classes
+// contain "impl" field that points to implementation object. "impl" field
+// should be accessible from this function.
+//
+// Note that due to a bug in MSVC compilers (including MSVC2019 v19.20), it
+// may not recognize the usage of this function in friend member declarations
+// if the template parameter name there is not equal to the name used here,
+// i.e. 'Obj'. For example, using 'Obj' here and 'T' in such declaration
+// would trigger that error in MSVC:
+//   template <class T>
+//   friend decltype(T::impl) detail::getSyclObjImpl(const T &SyclObject);
+template <class Obj> decltype(Obj::impl) getSyclObjImpl(const Obj &SyclObject) {
+  assert(SyclObject.impl && "every constructor should create an impl");
+  return SyclObject.impl;
+}
+
+// Returns the raw pointer to the impl object of given face object. The caller
+// must make sure the returned pointer is not captured in a field or otherwise
+// stored - i.e. must live only as on-stack value.
+template <class T>
+typename std::add_pointer_t<typename decltype(T::impl)::element_type>
+getRawSyclObjImpl(const T &SyclObject) {
+  return SyclObject.impl.get();
+}
```

Comment:
Do we really need this? It seems the only real use (per https://github.com/search?q=repo%3Aintel%2Fllvm%20getRawSyclObjImpl&type=code) is at https://github.com/intel/llvm/blob/7e7971a576665c8e0dfe5b985ada7cf1de58c5ce/sycl/source/detail/program_manager/program_manager.cpp#L117 which is on an exceptional path and we can afford an extra smart pointer copy.

Reply:
Can you explain your proposal? How do you get access to PIMPL context and device objects?

Reply:
Change uses like this: `getRawSyclObjImpl()` -> `getSyclObjImpl().get()`.

Reply:
@aelovikov-intel, I see many uses in "hot" path as well. _Do not assume that GitHub search results display all instances of the query!_

> we can afford an extra smart pointer copy.

@sergey-semenov, do you agree? IIRC, there were some optimizations in program manager, but I'm not sure if this proposal is related.

Reply:
The template was introduced by https://github.com/intel/llvm/commit/09a19d2088ff2ffb6d661fcaac62412a8ee27ce6, which is not related to performance optimization, but the performance concern is still present.
Meanwhile, I applied the suggestion and removed it in the https://github.com/intel/llvm/pull/10458/commits/b3539c565e50943f291fd244b9739ef2e7babf17. Let's see whether it passes CI checks.

Reply:
It seems `getRawSyclObjImpl()` and `getSyclObjImpl().get()` are not identical right now in terms of final asm, see https://godbolt.org/z/6sacE3hrc .
Suggest changing `getSyclObjImpl` to return by ref, then they get identical.

Reply:
`getSyclObjImpl` is much more widely used than `getRawSyclObjImpl`. I think that returning a reference requires accurate tracking of returned refence lifetime/ownership.

Reply:
I'll revert b3539c565e50 and create a separate issue to remove `getRawSyclObjImpl()`.